sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala

```diff
@@ -261,8 +261,16 @@ case class HashPartitioning(expressions: Seq[Expression], numPartitions: Int)
           expressions.length == h.expressions.length && expressions.zip(h.expressions).forall {
             case (l, r) => l.semanticEquals(r)
           }
-        case ClusteredDistribution(requiredClustering, _) =>
-          expressions.forall(x => requiredClustering.exists(_.semanticEquals(x)))
+        case c @ ClusteredDistribution(requiredClustering, _) =>
+          if (SQLConf.get.requireAllClusterKeysForHashPartition) {
```

Comment:
It seems to be worth to think of the trade-off on having a flag on ClusteredDistribution vs checking SQL config in here. For former, we may need to change the whole spots initializing ClusteredDistribution, but it is also meaning that we are open to finer-grained control in the future (additional mix-up conditions per-operator). For latter, it's probably the simplest change, but here we have no idea about the operator so we are restricted to apply the change in future as global manner.

I don't have strong preference on this as I'm talking about the extensibility which might not happen (or happen sooner). Just a 2 cents.

Reply:
I actually also thought about the pros and cons for these two approaches: 1). change behavior of `HashPartitioning`, vs 2). change behavior of `ClusteredDistribution` (for reference, [1] is an example of implementation). I am more inclined to 1). change behavior of `HashPartitioning` for followed reason:

`ClusteredDistribution`'s current definition is pretty clean and flexible, so let's not move backward.

``` scala
/**
 * Represents data where tuples that share the same values for the `clustering`
 * [[Expression Expressions]] will be co-located in the same partition.
 */
case class ClusteredDistribution
```

As long as data is partitioned in the way where tuples/rows having same values for `clustering` is in same partition, then the partitioning can satisfy `ClusteredDistribution`. It tolerates both full keys and subset of keys, so it's flexible enough to work for a range of operators - aggregate, window, join (together with `ShuffleSpec` introduced recently for co-partitioning). It does not has any implicit requirement of hash expression, or hash function (so it gets rid of the drawback of `HashClusteredDistribution`). More partitioning other than `HashPartitioning` can satisfy `ClusteredDistribution` (e.g. `RangePartitioning` and `DataSourcePartitioning`). Add flag such as `requiresFullKeysMatch` into `ClusteredDistribution` would make every `Partitioning` implementation unnecessarily more complicated, as this is just a problem for `HashPartitioning` now.

`HashPartitioning` can decide flexibly by itself when should it satisfy `ClusteredDistribution`, either subset of keys (current behavior), or full keys (with config introduced in this PR). This leaves other `Partitioning` (`RangePartitioning` and `DataSourcePartitioning`) and `ClusteredDistribution` untouched. Indeed this is just a local decision made by `HashPartitioning`. I think this is more flexible and extendable. In the future, if other `Partitioning` has similar requirement, e.g. `DataSourcePartitioning`, similar logic can be introduced inside `DataSourcePartitioning.satisfies0()` locally without any intrusive interface change.

> For latter, it's probably the simplest change, but here we have no idea about the operator so we are restricted to apply the change in future as global manner.

It's true, but I think the granularity is tricky to decide, so let's start with best solution to maintain our interface cleanly. We can discuss later if there is a strong requirement. One can further argue if user wants more finer granularity control  that he/she wants to specify exact operator in the query (e.g. the query has 3 `aggregate`s, and only wants to enable feature for 1 of them).

Reply:
I'm also inclined to option 1) for now, and agree to the points that @c21 raised above. 

As a Spark developer, I was originally confused when seeing both `HashClusteredDistribution` and `ClusteredDistribution` and had to navigate the code base and reason about their behavior differences. Combined with the newly introduced config, a developer now has to remember parsing the value of the config and choose `HashClusteredDistribution` or `ClusteredDistribution` accordingly, which is some extra burden. In addition, it's better to have a separate `StatefulOpClusteredDistribution` dedicated to SS use cases, as it makes them more distinctive.

Of course, having a separate `HashClusteredDistribution` opens up more opportunities for it to evolve separately. But I'd suggest to only consider that when we have some concrete ideas. So far, I don't see what can't be done with `ClusteredDistribution` alone.

Reply:
> As a Spark developer, I was originally confused when seeing both HashClusteredDistribution and ClusteredDistribution and had to navigate the code base and reason about their behavior differences.

The classdoc of two classes were very clear about differences. The confusion may come from the structure we have actual logic about requirement check for distribution in `partitioning` instead of distribution (I'm not an expert of this part, it might have to go with this way), but as long as the implementation matches up with classdoc, it is pretty clear.

> Combined with the newly introduced config, a developer now has to remember parsing the value of the config and choose HashClusteredDistribution or ClusteredDistribution accordingly, which is some extra burden.

Well, I'd say it is more extra burden if we have to expect two different requirements from ClusteredDistribution. Once we understand the difference between HashClusteredDistribution and ClusteredDistribution, it is obvious that we can easily infer the behavior from which class is used.

Reply:
We already made a single exception (`spark.sql.requireAllClusterKeysForCoPartition`), and we are going to make a broader exception here (`spark.sql.requireAllClusterKeysForHashPartition`).

There have been valid cases to support these exceptions, and they are not edge-cases like from odd data distribution. That said, this is going to be a valid requirement for distribution. In other words, there are known cases ClusteredDistribution solely does not solve the problem nicely. It is trying hard to eliminate shuffle as many as possible (assuming that shuffle is evil), but it is doing nothing for the cases shuffle does help.

So I don't think adding flag in ClusteredDistribution is messing up interface and structure. I think it is opposite. We are making exceptions for requirement of ClusteredDistribution - requiring full keys are not a one of requirements of ClusteredDistribution as of now, right? We haven't documented and it is now completely depending on the implementation of HashPartitioning. If someone starts to look into ClusteredDistribution, it is likely that someone misses the case. It is also possible we miss the config when implementing DataSourcePartitioning against ClusteredDistribution. I said "trade-off" because it pinpoints the issue and tries to bring a small amount of fix which may be preferable for someone, but my preference is making it clearer.

If we are skeptical to address this in ClusteredDistribution because we don't want to make requirement of ClusteredDistribution be mixed up, this is really a good rationalization we should revive back HashClusteredDistribution because the requirement is 100% fit to what we are doing. The only difference is that data source partitioning wouldn't satisfy the requirement in any way which may be considered as major downside according to the roadmap, so yes it brings data source partitioning as a second class for some cases. If we are against of it, please make sure ClusteredDistribution covers everything "by definition" HashClusteredDistribution could cover it solely.

Reply:
Maybe it's just me but the docs are not very crystal clear. First of all, the prefix `Hash` doesn't tell clearly the distinction between the two. Yes, it sounds like `HashClusteredDistribution` can only be satisfied by `HashPartitioning`, but how does it relates to the fact that Spark requires full partition keys match on `HashClusteredDistribution` while only requires partial match on `ClusteredDistribution`? Also, even though the doc says "This is a strictly stronger guarantee than ClusteredDistribution`, the "strict" here is a bit vague and I had to navigate to `HashPartitioning.satisfies` to find out the differences. There was also no context why `HashClusteredDistribution` is only used in joins while `ClusteredDistribution` is used in aggregates, etc.

> Well, I'd say it is more extra burden if we have to expect two different requirements from ClusteredDistribution. 

Again, think from Spark developers' point of view (for instance, someone who's creating a new operator). They have to first remember that these two configs exist, understand the differences between the two distributions and then write code to choose them accordingly, while in most cases it should just be transparent. This not only applies to developers on Spark itself but also those who're developing Spark extensions that replace Spark operators with their own.

By limiting the behavior only in config and `Partitioning` etc, the number of audiences affected by this change is much less:`Partitioning` is essentially a sealed trait so one will only be able to add new sub-classes within `partitioning.scala`. And yes, I'd certainly remember to check these for `DataSourcePartitioning` that I'm working on :)

Reply:
OK, 100% agree with that the class name does not tell the behavior very clear, especially there are both ClusteredDistribution and HashClusteredDistribution. In the other hands, the classdocs were very clear for both classes hence I didn't have to look into HashPartitioning.satisfy when figuring out the problem on required clustered distribution for stateful operator. But I agree everyone has a different view, so both can be right.

The concern comes from the fact that the unique functionality in HashClusteredDistribution is lost when unifying the classes. I'm not strongly concerning that we need to bring back HashClusteredDistribution. I'm concerning that requiring full keys is a valid requirement we shouldn't simply drop, at least I see from inputs.

> They have to first remember that these two configs exist, understand the differences between the two distributions and then write code to choose them accordingly, while in most cases it should just be transparent.

Seems like you'd like to say these are exceptional cases, but I'd wonder how we can say "exceptional cases" given we had to make exceptions two times in a couple of months, and the impact of latter is quite broader (not pinpointing the scope). I'd rather say it is no longer exceptional case.

The needs of requiring full grouping keys doesn't seem to be trivial. In the comment https://github.com/apache/spark/pull/35552#issuecomment-1045149556, @sigmod pointed out the valid case, and said AQE can't kick in because they are in the same stage. And the reason they are in the same stage is due of loose requirement of ClusteredDistribution. We can't simply prey AQE to deal with this nicely, at least we need to split out stages and ask AQE to handle the proper number of partitions.

What we make exceptions are all about the requirement of ClusteredDistribution. There is no problem for HashPartitioning. The reason we fix HashPartitioning is just due to the design that we use `satisfy`. Limiting the change to Partitioning and saying "We change nothing on ClusteredDistribution" doesn't seem right to me. We are technically changing ClusteredDistribution, and we need to make it clear. Otherwise this is also going to be an undocumented one. Instead of just documenting, I prefer having the flag explicitly so that any partitioning would know about the detailed requirement very clear. You don't need to remember the config when you work on DataSourcePartitioning. ClusteredDistribution will remember it for you.

Reply:
> I'm concerning that requiring full keys is a valid requirement we shouldn't simply drop, at least I see from inputs.

Just to make sure we are on the same page. The operators such as aggregate, window are always using `ClusteredDistribution` for years. So the problem of having data skew in those operators, is a new problem we are aiming to fix, not a regression coming from recent PRs.

> We are technically changing ClusteredDistribution, and we need to make it clear. Otherwise this is also going to be an undocumented one. Instead of just documenting, I prefer having the flag explicitly so that any partitioning would know about the detailed requirement very clear. You don't need to remember the config when you work on DataSourcePartitioning. ClusteredDistribution will provide it for you.

hmm maybe I am thinking too much from my perspective, but I am still not very convinced this is a problem for `ClusteredDistribution`. Hashing on subset of keys causes data skew seems to me a problem for `HashPartitioning` only. Other `Partitioning` such as `RangePartitioning` or `DataSourcePartitioning` can partition data very differently from `HashPartitioning`, or they do not use hash at all. So they might have very different causes other than subset of keys, to lead to data skew (e.g. suboptimal sampling algorithm for `RangePartitioning` to cause bad choice of partition boundary, or suboptimal user-defined `DataSourcePartitioning` to cause skew). I am kind of worried about introducing a flag such as `requiresFullKeysMatch` in `ClusteredDistribution` might be just useful for `HashPartitioning`, but not for other `Partitioning` classes. Once we introduce the flag, it's hard to change/remove the flag, because other developers or users depend on `DataSourcePartitioning` might be broken once we change `ClusteredDistribution` again. So just want to make sure we are very cautious about it.

Reply:
> Just to make sure we are on the same page. The operators such as aggregate, window are always using ClusteredDistribution for years. So the problem of having data skew in those operators, is a new problem we are aiming to fix, not a regression coming from recent PRs.

I agree with this, but the difference is that we dropped HashClusteredDistribution so we have less feature to leverage. In the SPIP doc we said "unify" two classes, but what we had done is "removing" the HashClusteredDistribution without alternatives.

Regarding the problem, there are two perspectives on the problem, 1) data skew 2) insufficient number of partitions. 2) applies to any partitioning.

Let's think about end user's perspective. They run the batch query, and expect Spark to finish it as quick as possible (or be resource-efficient). Spark produces the general config - default number of shuffle partitions - which defines the general parallelism whenever shuffle is introduced.

The point is when the config takes effect. More and more output partitioning ClusteredDistribution could match, less and less shuffle could be kicked in, while end users may expect like "a set of known operators would introduce shuffles, which adjusts parallelism as I set the config". For example, for the case ClusteredDistribution with DataSourcePartitioning, it could be hypothetically "nowhere". The max parallelism could be tied to source's parallelism and nothing except manual change of the query could help since there could be no shuffle at all. AQE won't help the parallelism/skew issue within the stage.

Shuffle may not be something we try hard to eliminate at all. We also need to think when the shuffle would be likely help. We can't leverage stats in physical execution, so my fall back goes to be heuristic, like the different view and resolution on this problem https://github.com/apache/spark/pull/35574#issuecomment-1047303366.

Reply:
In totally ideal world, previous stage could calculate the cardinality of all required grouping keys in next stage, and once the previous stage finishes, query executor decides to split the stage based on the difference of cardinality and the desired threshold of number of partitions (or even the number of values bound to each key). This is totally ideal and I don't know whether it is even possible. But if we agree that this is ideal, we are in agreement that shuffle is not always an evil.

Reply:
> Shuffle may not be something we should try hard to eliminate at all. We also need to think when the shuffle would be likely help. We can't leverage stats in physical execution, so my fall back goes to be heuristic, like the different view and resolution on this problem https://github.com/apache/spark/pull/35574#issuecomment-1047303366.

> Instead, having a threshold (minimum) of the number of partitions doesn't sound crazy for me. The threshold could be heuristic one, or config - number or ratio compared to the default number of shuffle partitions, or default number of shuffle partitions if we wouldn't want to bring another config (but it may be too high to use for minimum).

@HeartSaVioR - First I agree with you sometime shuffle is good to have, so I guess this PR is aiming for the same goal - add proper shuffle on full clustering keys based on config, right? Just for my understanding, are you proposing above to have a config to set the minimal threshold for number of partitions for all queries needed shuffle? Can you elaborate more how query and which part of query would be rewritten if violating the config? With cardinality stats from CBO and AQE (we don't have cardinality stats collected in AQE for now), we may potentially give some hint in query plan during logical planning. But this approach still sounds a little bit too high level for me, without elaborating details of algorithm.

Reply:
> I guess this PR is aiming for the same goal - add proper shuffle on full clustering keys based on config, right?

Yes, but we still have an argument that why it is needed. I wouldn't say it is due to HashPartitioning, as I mentioned about two different perspectives. Even if the hash function of HashPartitioning is somewhat inefficient on proper distribution on some sort of grouping keys and data, it is only contributing to data skew. In any partitioning, the number of partitions are the physical limit of parallelism. It is ClusteredDistribution preventing two operators to have shuffle in between, coupling operators in the same stage, with same partitioning & parallelism.
(Please correct me if AQE can decide to split the single stage to multiple stages injecting shuffles in between.)

Considering two different perspectives, there are multiple cases we may need to deal with:

1) clustered keys are fully considered and having sufficient number of partitions (ideal)
2) clustered keys are fully considered but having insufficient number of partitions
3) only a subset of clustered keys are considered and having insufficient number of partitions
4) only a subset of clustered keys are considered but having sufficient number of partitions

Since we are going with the manual / heuristic way to deal with it, I would like to see the way addressing more cases with less side-effects. That is the main reason I tried to think about alternatives.

Requiring full clustered keys can deal with 3) and 4), where 4) may be skewed (good case for shuffle) or not (shuffled isn't needed) which the benefit of having shuffle is conditional. Requiring a minimum threshold of number of partitioning can deal with 2) and 3) which are good in general to ensure minimum parallelism for grouping/joining operators, where it misses the case of 4), but we just mentioned the benefit of having shuffle in 4) is conditional. In addition, it is no longer only bound to data skew, hence applies to any partitioning.

Implementation wise, I imagine it is simple as we just add the another constraint of ClusteredDistribution. If the required number of partitions exists, it will strictly follow the number, otherwise we compare numPartitions of partitioning and the threshold being defined in ClusteredDistribution. The value of threshold could be optional if we doubt about the good default value working for majority of queries.